npm/pkg/dataplane/ipsets/ipsetmanager_linux.go

```diff
@@ -57,14 +60,135 @@ var (
+
+	example:
+		grep output:
+			azure-npm-123456
+			azure-npm-987654
+			azure-npm-777777
+
+		example restore file [flag meanings: -F (flush), -X (destroy)]:
+			-F azure-npm-123456
+			-F azure-npm-987654
+			-F azure-npm-777777
+			-X azure-npm-123456
+			-X azure-npm-987654
+			-X azure-npm-777777
+
+	prometheus metrics:
+		After this function, NumIPSets should be 0 or the number of NPM IPSets that existed and failed to be destroyed.
+		When NPM restarts, Prometheus metrics will initialize at 0, but NPM IPSets may exist.
+		We will reset ipset entry metrics if the restore succeeds whether or not some flushes/destroys failed (NOTE: this is different behavior than v1).
+		If a flush fails, we could update the num entries for that set, but that would be a lot of overhead.
+*/
 func (iMgr *IPSetManager) resetIPSets() error {
-	// called on failure or when NPM is created
-	// so no ipset cache. need to use ipset list like in ipsm.go
+	listCommand := iMgr.ioShim.Exec.Command(ipsetCommand, ipsetListFlag, ipsetNameFlag)
+	grepCommand := iMgr.ioShim.Exec.Command(ioutil.Grep, azureNPMPrefix)
+	searchResults, gotMatches, grepError := ioutil.PipeCommandToGrep(listCommand, grepCommand)
+	if grepError != nil {
+		return npmerrors.SimpleErrorWrapper("failed to run ipset list for resetting IPSets", grepError)
+	}
+	if !gotMatches {
+		metrics.ResetNumIPSets()
+		metrics.ResetIPSetEntries()
+		return nil
+	}
+	creator, originalNumSets, destroyFailureCount := iMgr.fileCreatorForReset(searchResults)
+	restoreError := creator.RunCommandWithFile(ipsetCommand, ipsetRestoreFlag)
+	if restoreError != nil {
+		metrics.SetNumIPSets(originalNumSets)
+		// NOTE: the num entries for sets may be incorrect if the restore fails
+		return npmerrors.SimpleErrorWrapper("failed to run ipset restore for resetting IPSets", restoreError)
+	}
+	if metrics.NumIPSetsIsPositive() {
+		metrics.SetNumIPSets(*destroyFailureCount)
+	} else {
+		metrics.ResetNumIPSets()
+	}
+	metrics.ResetIPSetEntries() // NOTE: the num entries for sets that fail to flush may be incorrect after this
+	return nil
+}
 
-	// create restore file that flushes all sets, then deletes all sets
-	// technically don't need to flush a hashset
+// this needs to be a separate function because we need to check creator contents in UTs
+func (iMgr *IPSetManager) fileCreatorForReset(ipsetListOutput []byte) (creator *ioutil.FileCreator, numSets int, destroyFailureCount *int) {
+	zero := 0
+	destroyFailureCount = &zero
+	creator = ioutil.NewFileCreator(iMgr.ioShim, maxTryCount, ipsetRestoreLineFailurePattern)
+	names := make([]string, 0)
+	readIndex := 0
+	var line []byte
+	// flush all the sets and create a list of the sets so we can destroy them
+	for readIndex < len(ipsetListOutput) {
+		line, readIndex = parse.Line(readIndex, ipsetListOutput)
+		hashedSetName := string(line)
+		names = append(names, hashedSetName)
+		// error handlers specific to resetting ipsets
+		errorHandlers := []*ioutil.LineErrorHandler{
```

Comment:
Great work to classify all errors and corresponding actions.
I understood the high-level ideas about `LineErrorHandler` which holds necessary error handling information to the corresponding executed line when there is error.

While I am fine as is, but just curious.
Do we consolidate all error handlings code without this `errorHandlers` here - https://github.com/Azure/azure-container-networking/blob/master/npm/pkg/dataplane/ioutil/restore_linux.go#L199?
For maintenance, it would be nice to consolidate error handling in one places if possible.
I thought `Callback` has only error message, but it changed `destroyFailureCount` in some places.

Reply:
Yep. I understood the intention of codes. 

My fundamental comments are whether we can consolidate all error definitions and handlers in one place per `ipset` and `iptables` instead of sparsely declaring them in each function and using `Callback`. 
It will be easier to understand and maintain codes, but I may miss somethings and it will need big changes.
wdyt?

BTW, it is fine as is.

Reply:
Still might not fully understand your question. To try to help consolidate, for `creatorForApply`, I group all of the error handling code at the bottom of the file. I can't move the error handling code out of `creatorForReset` because of the dependency on `destroyFailureCount`.

We could consolidate the definition and method (e.g. `ContinueAndAbortSection`) but the callback will always be unique and will eventually do more than logging for most (e.g. mark a set as needing reconcile).

Reply:
for instance though, we could make this error handler a function of the set name and destroyFailureCount pointer though